### ISSUE TITLE
Template functions request: strip domain, strip port

### ISSUE THREAD

**Issue by baryluk** [prometheus/prometheus#9932]
## Proposal
**Use case. Why is this important?**

Alerts should be concise, especially their summaries, yet many labels accessible during alert annotation construction will be long.

Example is `instance` or `target`, which usually will be a FQDN, often with port.

In our monitor system we also have a label `host` that is FQDN (it is used also for silencing). Often it is not derived from `instance`, but something else instead. This label is constructed in various ways using relabeling (often stripping port) or SD static labels.
But because we have about 30 different scrape jobs, there is a lot of repetitive relabeling rules to construct them.

Additionally to make alert summaries short, and help Grafana legend formating, and write alert silences quicker, we add `host_short`, which is `host` but only the host part, not FQDN. Again this is done using relabeling (with help of some regexp) or SD static labels for each target, depending on use case.

We use FQDNs for targets in SD, so DNS queries are correct, but 99% of the hosts are actually in the same domain, so there is no point of having the domain in alert annotations usually.

This approach has few downsides:

1) Stores more data in the TSDB and memory, which is arguably redundant data.

2) Relabeling rules for many targets are very repetitive and clutter the config file A LOT.

3) Grafana use case could be solved on Grafana side instead.

4) Because each time series now has `instance`, `host` and `host_short`, users are confused which one should be used for silence, or selectors, and `on` operators.

At the moment we do have 144 references to `host` and 102 `host_short` labels in our alerts, and it is just ugly.

The idea would be to not have these labels, but instead construct them during alert annotation rendering. However, this is even worse, and very tedious to do with just `reReplaceAll`. This could be workedaround by using pre-processor (in fact we use Python scripts to generate many alerting rules, for various reasons), but the generated yaml file as well viewing alerts definitions at `/alerts` will still be unreadable and very long.

So, I propose to add two helper string functions.

`stripPort` - removes colon and integer port parts after short or FQDN or IPv4 / IPv6 address, if any; and any trailing whitespaces

Example:

- "foo:123456" -> "foo"
- "foo" -> "foo"
- "foo.example.com:123456   " -> "foo.example.com"
- "foo.example.com." -> "foo.example.com."
- "127.0.0.1:33" -> "127.0.0.1"
- "127.0.0.1  " -> "127.0.0.1"
- "[::1]:9090" -> "[::1]"
- "[::1]" -> "[::1]"
- ":22" -> ""

`shortHostName` - if not IPv4 / IPv6, removes all the domain path elements, leaving only the first one; and any leading and trailing whitespaces in the output.

Example:

- "foo:123456" -> "foo"
- "  foo  " -> "foo"
- "foo.example.com:123456   " -> "foo"
- "foo.example.com." -> "foo"
- "127.0.0.1:33" -> "127.0.0.1"
- "127.0.0.1  " -> "127.0.0.1"
- "[::1]:9090" -> "[::1]"
- "[::1]" -> "[::1]"
- "" -> ""

Examples.

Instead of:

```yaml
...
# metric relabeling and target labels skipped
...
    - alert: NodeDown
      expr: 'up{job=~"node.*"} != 1'
      for: 2m
      labels:
        severity: critical
        dependency_level: machine
      annotations:
        summary: 'Network, machine or exporter issues for {{ $labels.host_short }}'
        description: '(location {{ $labels.location }}): Node exporter on {{ $labels.host }} is unreachable'
```

one could do this instead:

```yaml
...
# No metric relabeling or target labels
...

    - alert: NodeDown
      expr: 'up{job=~"node.*"} != 1'
      for: 2m
      labels:
        severity: critical
        dependency_level: machine
      annotations:
        summary: 'Network, machine or exporter issues for {{ $labels.instance | shortHostName }}'
        description: '(location {{ $labels.location }}): Node exporter on {{ $labels.instance | stripPort }} is unreachable'
```

More complex use cases, like extracting the domain name ("example.com") are rare, and specialized, that can be easily handled using `reReplaceAll` or metric relabeling.

**Comment by JessicaGreben** [comment]
Sounds good to implement `stripPort` and if there is an error return the original string, it should use https://pkg.go.dev/net#SplitHostPort.   If there is support in the go standard library (not inventing regex ourselves) for doing the desired logic in `shortHostName` then that sounds like a good idea too.

**Comment by uesyn** [comment]
This sounds like a good idea!
I would like to work on this issue if it's not assigned to anyone yet.
Can I grab this?

**Comment by roidelapluie** [comment]
Yes

**Comment by uesyn** [comment]
> stripPort - removes colon and integer port parts after short or FQDN or IPv4 / IPv6 address, if any; and any trailing whitespaces

Triming whitespaces is generic function, so I think `stripPort` should just do stripPort and simple port number check.
Therefore, I propose `trimSpace` in addition to `stripPort`.
I will implment it with https://pkg.go.dev/strings#TrimSpace.
So If you would like to trim whitespaces from host, you can do below:

```
{{ $labels.instance | stripPort | trimSpace }}
```

Do you think about it?

**Comment by baryluk** [comment]
> > stripPort - removes colon and integer port parts after short or FQDN or IPv4 / IPv6 address, if any; and any trailing whitespaces
> 
> Triming whitespaces is generic function, so I think `stripPort` should just do stripPort and simple port number check. Therefore, I propose `trimSpace` in addition to `stripPort`. I will implment it with https://pkg.go.dev/strings#TrimSpace. So If you would like to trim whitespaces from host, you can do below:
> 
> ```
> {{ $labels.instance | stripPort | trimSpace }}
> ```
> 
> Do you think about it?

I think this is better and less magic.

**Comment by peterbourgon** [comment]
```
{{ $labels.instance | reReplaceAll ":[0-9]*" "" }}
```

?

**Comment by uesyn** [comment]
@JessicaGreben 

Would you review this PR?
https://github.com/prometheus/prometheus/pull/10002

**Comment by ivange94** [comment]
@roidelapluie I see the PR linked to this issue is already merged. Is there something that still needs to be done here and is anyone working on it?

**Comment by binaryDiv** [comment]
The PR implemented `stripPort`, which is very nice and practical, thanks for that. :)

The other proposed function `shortHostName` is still open though.

I've never worked with Go until now, but it seems like an easy thing to implement. So unless anyone is already working on that, I could give it a try.

**Comment by roidelapluie** [comment]
Maybe we could find a better name .. stripDomain ?

**Comment by fatsheep9146** [comment]
"stripDomain" is just what we need in our case. If it's ok to you, I would like to implement this as soon as possible :） @binaryDiv

**Comment by peterbourgon** [comment]
Honest curiosity: do relabeling rules not solve this problem?

**Comment by roidelapluie** [comment]
Relabeling mangles the data before it's in TSDB. This is at the alerting level, e.g. to be used in templates. Those are 2 different levels.

**Comment by fatsheep9146** [comment]
@roidelapluie Would you review this pr? #10475

**Comment by baryluk** [comment]
Both proposed functions got implemented and released (in 2.33 for `stripPort` and 2.36 for `stripDomain`).

Big thanks to @uesyn  and @fatsheep9146
